### DISCUSSION TITLE
How is the performance of a GRPC collection server in Java? Can it scale according to the load?

### DISCUSSION THREAD

**Discussion by JangCool** [open-telemetry/opentelemetry-java-instrumentation#11885] [Q&A]
I'm currently looking into agent instrumentation and plan to examine the collection server in detail next. On GitHub, there are only simple examples for running a GRPC server, and I am curious about the performance differences compared to a Go-based collection server.

I read the document titled "Scaling the Collector," and it seems to recommend scaling under heavy load. Is it the same for a Java GRPC collection server? It might also be a good idea to separate the log collection server, tracing server, and metrics collection server.

[https://opentelemetry.io/docs/collector/scaling/](url)

I'm wondering whether it would be better to implement the exporter with Netty and reconfigure the collection server with Netty, or if a GRPC server would be sufficient..

**Comment by laurit** [comment]
Could you elaborate what do you mean by `Java GRPC collection server`?

**Reply by JangCool** [reply]
ah.. sorry.

Are you not involved with the opentelemetry-java-examples repository??

Since this is related to the same Java project, I thought you might be familiar with it, so I decided to ask here.

https://github.com/open-telemetry/opentelemetry-java-examples/tree/main/grpc

**Reply by laurit** [reply]
That example shows how to manually instrument grpc client and server. I still have no idea what you could mean with `Java GRPC collection server`.

**Reply by JangCool** [reply]
It seems there is an error in conveying the meaning.

I want to know if using the gRPC server as it is in the example will deliver good performance.

**Reply by laurit** [reply]
If performance is a concern I'd urge you to benchmark it yourself.

**Reply by trask** [reply]
https://github.com/open-telemetry/opentelemetry-java-examples/tree/main/grpc is not an otlp grpc server

**Reply by zerocooldog** [reply]
@trask Oh... I think there was some confusion.

I'm planning to implement a collection server using Java. I found an example of a gRPC server and wanted to ask if I can use that to implement the collection server.

Or, I was wondering if there are any plans to implement an OTEL collection server in Java language

**Reply by laurit** [reply]
For testing we use fake backend that accepts otlp/grpc https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/smoke-tests/images/fake-backend

**Reply by zerocooldog** [reply]
@laurit Thank you for the answer. I will take that into consideration.

And I think UDP communication might be necessary, so I should also look into implementing a custom exporter.
